### PR TITLE
fix: Incorrect promotion from index to paramater

### LINKED ISSUE #3458: Incorrect promotion from index to paramater
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

Some inductive types with indices get their indices promoted to parameters in cases where it shouldn't, leading to kernel-level errors afterwards

### Context

The issue was first discovered [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/bug.3A.20nonsensical.20disapperarence.20of.20kernel.20error.20message)

### Steps to Reproduce
MWE:
```lean
inductive All {I : Type u} (P : I → Type v) : List I → Type (max u v) where
  | cons : P x → All P xs → All P (x :: xs)

inductive Iμ {I : Type u}  : I → Type (max u v) where
  | mk : (i : I)  → All Iμ [] → Iμ i
```

**Expected behavior:** No error

**Actual behavior:** The index of `Iμ` gets promoted to a parameter during elaboration, leading to the following error:
```
arg #6 of '_nested.All_1.cons' contains a non valid occurrence
```

### Versions

nightly-2024-02-21
Ubuntu

### Additional Information

While there are some workarounds to avoid the promotion of indices to parameters, Lean would benefit from having an option to turn that feature off altogether.

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

### PR DESCRIPTION
**Pull request by arthur-adjedj**
Depends on #3590

Closes #3458

### CODE REVIEW

File: src/Lean/Elab/Inductive.lean [hunk 1]
```diff
@@ -25,6 +25,11 @@ open Meta
 builtin_initialize
   registerTraceClass `Elab.inductive
 
+register_builtin_option autoPromoteIndices : Bool := {
```
Comment (digama0) [inline src/Lean/Elab/Inductive.lean:28]:
```suggestion
register_builtin_option inductive.autoPromoteIndices : Bool := {
```
seems like it should be scoped in some way, although I'm really missing that option naming style guide :sob:

Comment (arthur-adjedj) [reply src/Lean/Elab/Inductive.lean]: I agree. Fixed in the relevant PR #3590.

**Comment by semorrison** [comment]
@arthur-adjedj, I've sent #3590 to the merge queue now. Please ping me when this one is ready to go.

**Comment by arthur-adjedj** [comment]
@semorrison This PR is in a ready state. My only worry is that the added condition to `visit` might be a bit costly, since the function is called quite frequently in other places. If that was the case, it might be better to make another function entirely for the `stopWhenVisited := true` case. Could this warrant a `!bench` ?

**Comment by semorrison** [comment]
!bench

File: src/Lean/Util/ForEachExprWhere.lean [hunk 2]
```diff
@@ -58,14 +58,16 @@ def checked (e : Expr) : ForEachM m Bool := do
     return false
 
 /-- `Expr.forEachWhere` (unsafe) implementation -/
-unsafe def visit (p : Expr → Bool) (f : Expr → m Unit) (e : Expr) : m Unit := do
+unsafe def visit (p : Expr → Bool) (f : Expr → m Unit) (e : Expr) (stopWhenVisited : Bool := false): m Unit := do
```
Comment (semorrison) [inline src/Lean/Util/ForEachExprWhere.lean:61]:
```suggestion
unsafe def visit (p : Expr → Bool) (f : Expr → m Unit) (e : Expr) (stopWhenVisited : Bool := false) : m Unit := do
```

File: src/Lean/Util/ForEachExprWhere.lean [hunk 3]
```diff
@@ -78,9 +80,11 @@ where
...
 
 /--
-`e.forEachWhere p f` applies `f` to each subterm that satisfies `p`.
+  `e.forEachWhere p f` applies `f` to each subterm that satisfies `p`.
+  If `stopWhenVisited` is `true`, the function doesn't visit subterms of terms
+  which satisfy `p`.
 -/
 @[implemented_by ForEachExprWhere.visit]
-opaque Expr.forEachWhere {ω : Type} {m : Type → Type} [STWorld ω m] [MonadLiftT (ST ω) m] [Monad m] (p : Expr → Bool) (f : Expr → m Unit) (e : Expr) : m Unit
+opaque Expr.forEachWhere {ω : Type} {m : Type → Type} [STWorld ω m] [MonadLiftT (ST ω) m] [Monad m] (p : Expr → Bool) (f : Expr → m Unit) (e : Expr) (stopWhenVisited : Bool := false): m Unit
```
Comment (semorrison) [inline src/Lean/Util/ForEachExprWhere.lean:88]:
```suggestion
opaque Expr.forEachWhere {ω : Type} {m : Type → Type} [STWorld ω m] [MonadLiftT (ST ω) m] [Monad m] (p : Expr → Bool) (f : Expr → m Unit) (e : Expr) (stopWhenVisited : Bool := false) : m Unit
```

File: tests/lean/run/3458_2.lean [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/- Ensure that non-fixed indices don't get mistakenly promoted to parameter
+   See issue `#3458`
```
Comment (semorrison) [inline tests/lean/run/3458_2.lean:2]: Let's use the whole URL so people can just click.

**Comment by semorrison** [comment]
Sorry it took a while to get back to this.

**Pull request merged**